### PR TITLE
text/openFile method

### LINKED ISSUE #546: Implement `text/openFile`
### Summary
For text files that are able to be edited collaboratively it's very important that the language server knows which of the files are open by which clients. The `text/openFile` message is what informs the LS about this.

### Value
Knowing which files are open is incredibly important for the purposes of both updating clients on changes and knowing when to read files from disk or from an in-memory representation. This message ensures that the language server can retain a coherent picture of the project. 

### Specification

- [ ] Add a message `workspace/connect` that lets new clients connect and identify themselves.
- [ ] Implement the `text/openFile` message as specified in the [design document](https://github.com/luna/enso/blob/master/doc/design/engine/engine-services.md).
- [ ] This should track which files are open on a per-client basis, and keep them loaded into in-memory buffers.
- [ ] Specify any errors in the design document.

### Acceptance Criteria & Test Cases

- The above specification has been implemented.
- The features have been rigorously tested.

### PR DESCRIPTION
**Pull request by lolczak**
### Pull Request Description
Implements the text/openFile method for the language server.
Automatically handles granting the proper capability on opening the file.

Resolves #546.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/WebSocketServer.scala [hunk 1]
```diff
@@ -60,15 +63,24 @@ class WebSocketServer(
...
+      )
 
     val messageHandler =
       system.actorOf(
         Props(new MessageHandler(ClientApi.protocol, clientActor))
       )
     clientActor ! ClientApi.WebConnect(messageHandler)
 
-    languageServer ! LanguageProtocol.Connect(clientId, clientActor)
+    system.eventStream.publish(ClientConnected(Client(clientId, clientActor)))
```
Comment (kustosz) [inline engine/language-server/src/main/scala/org/enso/languageserver/WebSocketServer.scala]: Just a thought – wouldn't this look better sent from `clientActor` upon it receiving the `WebConnect`? Would be more consistent, given that the client actor the one sending `disconnect` as well.

File: engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/AcquireCapabilityHandler.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    case RequestTimeout =>
+      log.error(s"Acquiring capability for ${client.id} timed out")
+      replyTo ! ResponseError(Some(id), ServiceError)
+      context.stop(self)
+
+    case CapabilityAcquired =>
+      replyTo ! ResponseResult(AcquireCapability, id, Unused)
+      context.stop(self)
+
+    case CapabilityAcquisitionBadRequest =>
```
[the remark this answers is not in the document]
Comment (kustosz) [reply engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/AcquireCapabilityHandler.scala]: (that concerns all response handlers, not just this one)

File: engine/language-server/src/main/scala/org/enso/languageserver/WebSocketServer.scala (hunk 1 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/WebSocketServer.scala]: good point

File: engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/AcquireCapabilityHandler.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/requesthandler/AcquireCapabilityHandler.scala]: fixed

File: engine/language-server/src/main/scala/org/enso/languageserver/LanguageServer.scala [hunk 3]
```diff
@@ -12,50 +17,64 @@ object LanguageProtocol {
...
-    *
-    * @param clientId the client to grant the capability to.
-    * @param registration the capability to grant.
+    * Confirms client that capability has been acquired.
     */
-  case class AcquireCapability(
-    clientId: Client.Id,
-    registration: CapabilityRegistration
-  )
+  case object CapabilityAcquired extends AcquireCapabilityResponse
```
[the remark this answers is not in the document]
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/LanguageServer.scala]: fixed

File: engine/language-server/src/main/scala/org/enso/languageserver/WebSocketServer.scala (hunk 1 above)
Comment (lolczak) [reply engine/language-server/src/main/scala/org/enso/languageserver/WebSocketServer.scala]: fixed

File: build.sbt [hunk 4]
```diff
@@ -544,6 +545,10 @@ lazy val runner = project
     assemblyJarName in assembly := "enso.jar",
     test in assembly := {},
     assemblyOutputPath in assembly := file("enso.jar"),
+    assemblyMergeStrategy in assembly := {
```
[the remark this answers is not in the document]
Comment (lolczak) [reply build.sbt]: I adjusted it and the LangServ works.

[the remark this answers is not in the document]
Comment (kustosz) [reply build.sbt]: But `--repl` does not, just tested. The META-INF used by Graal to run our language is missing.

**Pull request merged**